tree/ntuple/v7/src/RNTupleIndex.cxx

```diff
@@ -115,31 +129,45 @@ void ROOT::Experimental::Internal::RNTupleIndex::Build()
 ROOT::Experimental::NTupleSize_t
 ROOT::Experimental::Internal::RNTupleIndex::GetFirstEntryNumber(const std::vector<void *> &valuePtrs) const
 {
-   const auto entryIndices = GetAllEntryNumbers(valuePtrs);
-   if (!entryIndices)
+   const auto entryNumbers = GetAllEntryNumbers(valuePtrs);
+   if (entryNumbers.empty())
       return kInvalidNTupleIndex;
-   return entryIndices->front();
+   return entryNumbers.front();
 }
 
-const std::vector<ROOT::Experimental::NTupleSize_t> *
+const std::vector<ROOT::Experimental::NTupleSize_t>
 ROOT::Experimental::Internal::RNTupleIndex::GetAllEntryNumbers(const std::vector<void *> &valuePtrs) const
 {
    if (valuePtrs.size() != fIndexFields.size())
       throw RException(R__FAIL("Number of value pointers must match number of indexed fields."));
 
    EnsureBuilt();
 
-   std::vector<NTupleIndexValue_t> indexValues;
-   indexValues.reserve(fIndexFields.size());
+   std::vector<std::vector<NTupleIndexValue_t>> entryNumbersPerCluster;
+
+   std::vector<NTupleIndexValue_t> indexFieldValues;
+   indexFieldValues.reserve(fIndexFields.size());
 
    for (unsigned i = 0; i < valuePtrs.size(); ++i) {
-      indexValues.push_back(CastValuePtr(valuePtrs[i], *fIndexFields[i]));
+      indexFieldValues.push_back(CastValuePtr(valuePtrs[i], *fIndexFields[i]));
    }
 
-   auto entryNumber = fIndex.find(RIndexValue(indexValues));
+   RIndexValue indexValue(indexFieldValues);
 
-   if (entryNumber == fIndex.end())
-      return nullptr;
+   for (const auto &indexPartition : fIndexPartitions) {
+      auto clusterEntryNumbers = indexPartition.fIndex.find(indexValue);
+
+      if (clusterEntryNumbers == indexPartition.fIndex.end())
+         continue;
+
+      entryNumbersPerCluster.push_back(clusterEntryNumbers->second);
+   }
+
+   std::vector<NTupleIndexValue_t> entryNumbers;
+
+   for (const auto &clusterEntries : entryNumbersPerCluster) {
+      entryNumbers.insert(entryNumbers.end(), clusterEntries.cbegin(), clusterEntries.cend());
+   }
 
-   return &(entryNumber->second);
+   return entryNumbers;
```

Comment:
(side note: it might be possible to get around returning vectors by value. For example, after building the partitions the `RNTupleIndex` could "link" collisions across partitions. Then we could return a "linked list of vectors", which would behave like a container with a custom iterator.)